### ISSUE TITLE
fix the flakes in "kind export logs"

### ISSUE THREAD

**Issue by neolit123** [kubernetes-sigs/kind#456]
**What happened**:
`kind export logs` can flake on very rare occasions.
me and @BenTheElder have seen this in CI before and it still happens.

**What you expected to happen**:
no flakes.

**How to reproduce it (as minimally and precisely as possible)**:
- create a 1CP2W cluster using `kind create cluster`
- write this to a bash script:

```bash
#!/bin/bash
TMPDIR="$(mktemp -d)"
for i in {1..1000}; do
    echo "----------------- $i"
    kind export logs "${TMPDIR}" || exit 1
    rm -rf "${TMPDIR}/*.*"
done
```
- call this script and check on it periodically (it will run for a while...)
- the command will fail on very rare occasions
- if i extend the logging a bit i'm seeing:
```
Error: error in execToPath for: kind-control-plane/pods/0820b847-67a1-11e9-b78f-0242c21cba65/coredns/0.log: exit status 126
```
or
```
Error: error in execToPath for: kind-worker2/kubernetes-version.txt: exit status 126
```

i think the 126 comes from:
https://github.com/kubernetes-sigs/kind/blob/master/pkg/cluster/logs/logs.go#L113
but not sure why, possibly a file gets an exclusive lock for a period of time.

**Anything else we need to know?**:
no

**Environment:**
any?

- kind version: latest master
- Kubernetes version: does not matter
- Docker version: doesn't seem to matter
- OS doesn't seem to matter

/assign
/kind bug
/priority important-longterm

**Comment by aojea** [comment]
googling the error exit status 126 says
>  126 as exit status which means "Command was found but could not be executed".
it can be possible that the errors is because we are not using the full path in the commands?

**Comment by neolit123** [comment]
126 can also be related to permissions.
https://www.tldp.org/LDP/abs/html/exitcodes.html

> Permission problem or command is not an executable

**Comment by BenTheElder** [comment]
https://docs.docker.com/engine/reference/run/#/exit-status

**Comment by aojea** [comment]
I´m trying to reproduce it without success, my observation is that `kind` runs in parallel several goroutines 
```
21802 pts/4    Sl+    0:00  |       \_ ./kind export logs /tmp/tmp.KIXat1dxEE
22151 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/pods/514670304442e821ba1e3d444ab0a987/kube-apiserver/0.log
22156 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/pods/1b8f1d356ff49cde2492b047b72d1b56/etcd/0.log
22157 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/containers/etcd-kind-control-plane_kube-system_etcd-024c79302b261
22158 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/pods/ac8420a7-684e-11e9-82bb-02424a48f45e/kube-proxy/0.log
22159 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/containers/weave-net-l5hsh_kube-system_weave-npc-ce5074d19d7eef33
22160 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/pods/acd871c3-684e-11e9-82bb-02424a48f45e/weave-npc/0.log
22162 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/containers/kube-apiserver-kind-control-plane_kube-system_kube-api
22163 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/pods/721a27dac190bee34730703b3377eca9/kube-controller-manager/0.l
22192 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/containers/kube-controller-manager-kind-control-plane_kube-system
22219 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/pods/4b52d75cab61380f07c0c5a69fb371d4/kube-scheduler/0.log
22253 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/containers/weave-net-l5hsh_kube-system_weave-d5e61db6dccfe28e8823
22258 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/containers/kube-scheduler-kind-control-plane_kube-system_kube-sch
22280 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/containers/kube-proxy-jk8xw_kube-system_kube-proxy-d09aee44a1bd4e
22298 pts/4    Sl+    0:00  |           \_ docker exec --privileged -t kind-control-plane cat /var/log/pods/acd871c3-684e-11e9-82bb-02424a48f45e/weave/0.log
```

do we know what command is throwing the exit status 126? is the cat inside the container?

**Comment by BenTheElder** [comment]
yes we collect logs in parallel.

cat should only be exiting 1 or 0 from a quick skim of it https://git.savannah.gnu.org/cgit/coreutils.git/plain/src/cat.c

126 should be `docker exec`, see the exit codes listing above

**Comment by neolit123** [comment]
> cat should only be exiting 1 or 0 from a quick skim of it https://git.savannah.gnu.org/cgit/coreutils.git/plain/src/cat.c

seems like so.

**Comment by neolit123** [comment]
could be docker version related - i'm at 18.06.x.
i'm going to bake a test that will spam `docker exec` many times to see if this manifests again.

also wonder if it can happen during parallel node creation.

**Comment by aojea** [comment]
can´t hit it with 
`parallel -j50 -N0 docker exec --privileged -t kind-control-plane cat /kind/version ::: {1..1000}`

and 
```
docker version
Client:
 Version:           18.09.0
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        e68fc7a215d7
 Built:             Fri Jan 11 12:00:00 2019
 OS/Arch:           linux/amd64
 Experimental:      false
```

**Comment by neolit123** [comment]
yes, i'm hitting it in a lot of different ways.

```
parallel -j50 -N0 docker exec --privileged -t kind-control-plane cat /kind/version ::: {1..1000}
````
^ this (or cat on a single container) passes most of the time.

but it failed for me with something like (e.g. working on two containers with more logs to cat):
```
parallel -j50 -N0 --halt now,fa=1 "docker exec --privileged -t kind-control-plane cat /kind/version > /dev/null && docker exec --privileged -t kind-worker cat /var/log/containers/weave-net-5v7t4_kube-system_weave-183bc46eba1443c9fd0bdfcfc59a0bfcf907de48f8f2687cd1915f163c1fde1c.log > /dev/null" ::: {1..1000}
```

it's a `docker exec` problem...and is rare.
question is how should we mitigate it.

**Comment by neolit123** [comment]
@BenTheElder any objections to adding the optional `kind export logs --serial`?

**Comment by BenTheElder** [comment]
probably just adding a retry

**Comment by aojea** [comment]
@neolit123 it seems that those `docker exec` are `POST` requests, if I check the journalctl I have a lot of errors, maybe is a problem of tweaking the OS

```
Apr 26 23:28:38 linux-6my5 dockerd[2201]: time="2019-04-26T23:28:38.893148693+02:00" level=error msg="Handler for GET /v1.39/conta
iners/kind-control-plane/json returned error: write unix /var/run/docker.sock->@: write: broken pipe"
```

what about `docker cp`?

**Comment by BenTheElder** [comment]
we can't use `docker cp` for some of these, check the source for details but TLDR we need to walk symlinks.

> @BenTheElder any objections to adding the optional kind export logs --serial?

a small one, the fact that it is parallel shouldn't be exposed, we should fix the reliability issue.

if we fail we can retry once.

**Comment by neolit123** [comment]
i'm also catching instances where the `export logs` call blocked indefinitely.
this cluster has been running for a few hours.

> if we fail we can retry once.

something in these lines?
```
@@ -133,7 +135,14 @@ func coalesce(fns ...errFn) error {
                wg.Add(1)
                go func(f errFn) {
                        defer wg.Done()
-                       ch <- f()
+                       // do a single retry in case of an error
+                       // https://github.com/kubernetes-sigs/kind/issues/456
+                       err := f()
+                       if err != nil {
+                               err = f()
+                       }
+                       ch <- err
                }(fn)
```

**Comment by aojea** [comment]
The command `docker exec kind-control-plane tar Chcf /var/log - . | tar Cxf /tmp/test/logs -` does the trick to move all the `/var/log` folder in one stream.

I think that we should use it to avoid contention on the docker exec command because it doesn´t seem to scale well https://github.com/kubernetes-sigs/kind/issues/456#issuecomment-487222334

**Comment by neolit123** [comment]
+1 ^

**Comment by aojea** [comment]
I can´t reproduce the errors anymore with #458 , anyway appreciate if others can confirm

**Comment by BenTheElder** [comment]
+1 for avoiding contention this way FWIW :^)

**Comment by BenTheElder** [comment]
I think #475 should have this fixed (which is basically #458, thanks @aojea !!)
